bindings/pyroot_experimental/PyROOT/python/ROOT/pythonization/_ttree.py

```diff
@@ -31,15 +33,19 @@ def pythonize_ttree(klass, name):
     # klass: class to be pythonized
     # name: string containing the name of the class
 
-    if name == 'TTree':
+    if TClass.GetClass(name).InheritsFrom('TTree'):
```

Comment:
How expensive is this call? Does it invoke the interpreter?

Reply:
I'm pretty sure it does ;) But does it slow down the invokation of the pythonizations significantly? We call this for each new class touched in Python!

Reply:
We need desperately a benchmark checking the invocation of the pythonizations!

Reply:
Thanks for pointing this out @stwunsch , indeed this is potentially more costly than just plain string comparison. I will compare the performance of both with a simple benchmark.

The alternative to this would be to have an explicit list with the class names and check that list, but that would be less flexible because it would not cover any subclass of `TTree` (users might even define their own subclasses).

Reply:
In any case, since there are a few cases of common pythonizations (`TDirectory` and children are another example), for the sake of avoiding code duplication I believe it is good to group them somehow.

Reply:
Another optimisation that we could apply here is the deregistration of pythonizor functions. When a pythonizor realizes that it has fulfilled its mission, it can deregister itself from the list of pythonizor (the API of `cppyy` has a function for that).

Reply:
An improvement on the above code would be to use the python equivalent of:
```
static TClass *treeClass = TClass::GetClass("TTree");
if ( TClass::GetClass("name")->InheritsFrom(treeClass) ) ..
```
of maybe even just
```
if TClass.GetClass(name).InheritsFrom(TTree.Class()):
```
To avoid the lookup of the TTree class.

Reply:
Hi @pcanal thanks for the suggestion, that would improve the solution but we would still have the issue that we would touch `libTree` even if the user Python script never uses anything from that library, since the pythonizor for `TTree` will be fired at first touch of **any** class. Comparing strings, even if less flexible, avoids that problem.

Reply:
> string ... but that would be less flexible because it would not cover any subclass of TTree (users might even define their own subclasses).

So the cost of 'not loading libTree' is that user derived classes are never pythonized ...  So the real question is where the performance gain is worth the lost in functionality ....

Reply:
> So the cost of 'not loading libTree' is that user derived classes are never pythonized ... So the real question is where the performance gain is worth the lost in functionality ....

Yes, that is the question. However, we intend to provide means for users to define pythonizor functions for their classes. So even if ROOT pythonizors would only target ROOT classes, user pythonizors could still explicitly inject any desired behaviour (perhaps reusing ROOT pythonizations).

Reply:
Also the execution of the 'fast' code could be predictated with a check on whether libTree has already been loaded or not.

Reply:
> Also the execution of the 'fast' code could be predictated with a check on whether libTree has already been loaded or not.

Wouldn't that mean that a class `A`, user subclass of `TTree`, would be pythonized or not depending on whether `libTree` has already been loaded or not? Since the fast code is just a check on a list of ROOT classes, if `A` were used after `TTree`, it would be pythonized (slow check mode based on inheritance), if it were used before, it would not be (fast check based on list).

Reply:
If class A inherits from class TTree then libTree **must** be loaded in order for class A to even come into existence (or it will have missing symbols including but not limited to the constructor and destructor of TTree that are referenced by class A's constructor and destructor )

Reply:
Yes, but if I understood correctly you are proposing to check if `libTree` has been loaded to decide if we take the fast path, so this (pseudocode):
```python
if libTree_is_not_loaded:
    if A_class in list_of_classes:   # fast check
        pythonize(A_class)
else:
    if A_class.InheritsFrom('TTree'): # slow check
        pythonize(A_class)
```
Note that if `libTree` has not been loaded, if `A_class` is a user class that is subclass of `TTree`, it will not be pythonized (because it is not on the list). It will be pythonized only if `libTree` was loaded (triggered by one of the classes in the list).

Reply:
> so this (pseudocode):

Yes that is what I meant .... (except for using the 'faster' version of the 'slow check')

> Note that if libTree has not been loaded, if A_class is a user class that is subclass of TTree

And  my claims is that this can not happen.  I.e. if the pythonization is being passed a class deriving from TTree, the libTree **must** have been already loaded (otherwise how did you get an object of that type ....)   

And actually I think this is true even of TTree itself.  So I think the logic can be simplified down to:

```
if libTree_is_loaded:
    if A_class.InheritsFrom(TTree.Class()):
        pythonize(A_class)
```

Reply:
Yes, that would work, my only worry is again the cost. I guess `if libTree_is_loaded` would mean calling `TInterpreter::IsLibraryLoaded`, which might involve a fair amount of string comparisons. In addition, we would pay that cost for every new class that is touched from the user script, since it is a top-level check in the pythonizor. Also, if we implement the same strategy for e.g. the pyhonizations of `TDirectory`&children, we would pay the cost again in that other pythonizor.

What I would suggest is to leave for now the fast check on a fixed list of ROOT TTree subclasses we want to pythonize, and if the fact that user subclasses of ROOT classes are not automatically pythonized by ROOT becomes an issue, we know we have this other strategy.

Thanks for the talk, this was useful!